<div class="alert alert-info">
Привет, Сергей! Меня зовут Светлана Чих и я буду проверять твой проект. Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе. Предлагаю общаться на «ты». Но если это не удобно - дай знать, и мы перейдем на «вы».

<div class="alert alert-success">
<b>👍 Успех:</b> Зелёным цветом отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах.
</div>
<div class="alert alert-warning">
<b>🤔 Рекомендация:</b> Жёлтым цветом выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).
</div>
<div class="alert alert-danger">
<b>😔 Необходимо исправить:</b> Красным цветом выделены комментарии, без исправления которых, я не смогу принять проект :(
</div>
<div class="alert alert-info">
<b>👂 Совет:</b> Какие-то дополнительные материалы
</div>
Давай работать над проектом в диалоге: если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом.
Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Например, вот так.</div>
Пожалуйста, не перемещай, не изменяй и не удаляй мои комментарии. Всё это поможет выполнить повторную проверку твоего проекта быстрее.
 </div>


# Описание проекта SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Описание данных

**Таблица books
Содержит данные о книгах:**

•	book_id — идентификатор книги;  
•	author_id — идентификатор автора;  
•	title — название книги;  
•	num_pages — количество страниц;  
•	publication_date — дата публикации книги;  
•	publisher_id — идентификатор издателя.


**Таблица authors
Содержит данные об авторах:**

•	author_id — идентификатор автора;  
•	author — имя автора.


**Таблица publishers
Содержит данные об издательствах:**

•	publisher_id — идентификатор издательства;  
•	publisher — название издательства;


**Таблица ratings
Содержит данные о пользовательских оценках книг:**

•	rating_id — идентификатор оценки;  
•	book_id — идентификатор книги;  
•	username — имя пользователя, оставившего оценку;  
•	rating — оценка книги.


**Таблица reviews
Содержит данные о пользовательских обзорах на книги:**

•	review_id — идентификатор обзора;  
•	book_id — идентификатор книги;  
•	username — имя пользователя, написавшего обзор;  
•	text — текст обзора.


## План выполнения задания

**Задания**

•	Посчитайте, сколько книг вышло после 1 января 2000 года;  
•	Для каждой книги посчитайте количество обзоров и среднюю оценку;  
•   Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;  
•	Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;  
•   Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.


**План выполнения?**

•	Опишите цели исследования;  
•	Исследуйте таблицы — выведите первые строки;  
•	Сделайте по одному SQL-запросу для решения каждого задания;  
•	Выведите результат каждого запроса в тетрадке;  
•	Опишите выводы по каждой из решённых задач.


In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [2]:
# создадим функцию для вывода результата запроса
def result(query):
    return pd.io.sql.read_sql(query, con = engine)


In [3]:
# выведем первые 10 строк таблицы
query ='''

SELECT *
FROM books
LIMIT 10 

'''
result(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

 <div class="alert alert-success">
 <b>👍 Успех:</b> Есть описание задачи, импортированы все нужные библиотеки, создано подключение к БД, просмотрены таблицы
 </div>

### Задание 1.
Посчитайте, сколько книг вышло после 1 января 2000 года;

In [4]:
query ='''

SELECT COUNT (book_id)
FROM books
WHERE publication_date > '2000-01-01'

'''
result(query)

count
0    819

Вывод: 819 книг вышло после 2000-01-01

<div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

### Задание 2.
Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [19]:
query ='''

SELECT b.book_id,
       b.title,
       rat.avg_rating,
       rev.review_count
FROM books b
LEFT JOIN (SELECT book_id, AVG(rating) avg_rating FROM ratings GROUP BY book_id) rat ON b.book_id=rat.book_id
LEFT JOIN (SELECT book_id, COUNT(*) review_count FROM reviews GROUP BY book_id) rev ON b.book_id=rev.book_id



'''
result(query)

book_id                                              title  avg_rating  \
0        652          The Body in the Library (Miss Marple  #3)    4.500000   
1        273                                          Galápagos    4.500000   
2         51                           A Tree Grows in Brooklyn    4.250000   
3        951  Undaunted Courage: The Pioneering First Missio...    4.000000   
4        839                                        The Prophet    4.285714   
..       ...                                                ...         ...   
995       64                                Alice in Wonderland    4.230769   
996       55          A Woman of Substance (Emma Harte Saga #1)    5.000000   
997      148                                          Christine    3.428571   
998      790  The Magicians' Guild (Black Magician Trilogy  #1)    3.500000   
999      828                           The Plot Against America    3.000000   

     review_count  
0             2.0  
1             2.0  
2             5.0  
3             2.0  
4             4.0  
..            ...  
995           4.0  
996           2.0  
997           3.0  
998           2.0  
999           2.0  

[1000 rows x 4 columns]

Вывод: Twilight кажется настоящий блокбастер, 1120 отзывов. В 2 раза больше,чем у ближайшего конкурента The Hobbit or  There and Back Again

 <div class="alert alert-danger">
<s> <b>😔 Необходимо исправить:</b> В итоговую таблицу вошли не все книги, а вот обзоров получилось слишком много, это связано с особенностью объединения таблиц и некоторые посчиталсь по несколько раз</s>
 </div>

<div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Запрос испрвил. Тепрь вроде похоже на правду.</div>

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

### Задача 3.
Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [6]:
query ='''

SELECT p.publisher_id,
       publisher,
       COUNT (b.book_id) as count     
FROM books b
JOIN publishers p ON p.publisher_id = b.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher_id
ORDER BY count DESC
LIMIT 1


'''
result(query)

publisher_id      publisher  count
0           212  Penguin Books     42

Вывод: Издательство Penguin Books издает больше всего книг, толще 50 стр., 42 штуки.

<div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

### Задача 4.
•	Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [7]:
query ='''

SELECT 
       ath.author,
       ROUND (AVG (r.rating),2) AS avg_rat,
       COUNT (r.rating) As cnt_rat
FROM authors AS ath
JOIN books b ON ath.author_id = b.author_id
JOIN ratings r on r.book_id = b.book_id
WHERE b.book_id IN 
                  (SELECT r.book_id      
                   FROM ratings r  
                   GROUP BY r.book_id
                   HAVING COUNT (r.book_id) >= 50)
GROUP BY ath.author
ORDER BY avg_rat DESC
LIMIT 1

'''
result(query)

author  avg_rat  cnt_rat
0  J.K. Rowling/Mary GrandPré     4.29      310

Вывод: J.K. Rowling/Mary GrandPré имеет наивысший рейтинг произведений - 4.29. Всего оценок -	310

<div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

### Задача 5.
Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [8]:
query ='''
WITH i AS

(SELECT COUNT(review_id) cnt, 
        username
FROM reviews 
GROUP BY username 
HAVING username IN (
                     SELECT username
                     FROM ratings
                     GROUP BY username 
                     HAVING COUNT(rating_id)>50
                     ) )
      
SELECT ROUND(AVG (i.cnt),2)
FROM i


'''
result(query)

round
0  24.33

Вывод: среднее количество обзоров от пользователей, которые поставили больше 50 оценок = 24 обзора

<div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

# Общие выводы 
•	Посчитайте, сколько книг вышло после 1 января 2000 года;  
_819 книг вышло после 2000-01-01_  
•	Для каждой книги посчитайте количество обзоров и среднюю оценку;  
_Twilight кажется настоящий блокбастер, 1120 отзывов. В 2 раза больше,чем у ближайшего конкурента The Hobbit or There and Back Again_  
•   Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;  
_Издательство Penguin Books издает больше всего книг, толще 50 стр., 42 штуки._  
•	Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;  
_J.K. Rowling/Mary GrandPré имеет наивысший рейтинг произведений - 4.29. Всего оценок - 310_  
•	Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.
_Среднее количество обзоров от пользователей, которые поставили больше 50 оценок = 24 обзора_


 <div class="alert alert-success">
 <b>👍 Успех:</b> Отличная работа! Поздравляю с завершением курса! Теперь ты настоящий аналитик))) Если захочешь расширить свои знания SQL, то можно посмотреть вот <a href='https://stepik.org/course/3203/promo'> сюда</a>
 </div>